### PR TITLE
feat(nuget): Support ranges

### LINKED ISSUE #5016: Add full range support to nuget versionScheme
**What would you like Renovate to be able to do?**

Add full support for NuGet's range/constraint syntax in lib/versioning/nuget

**Describe the solution you'd like**

Should be similar to other version scheme range support, e.g. Maven

**Describe alternatives you've considered**

The support added in #4947 is a reasonable workaround

**Additional context**

See discussion in #4947

### LINKED ISSUE #7191: Support Nuget ranges
**What would you like Renovate to be able to do?**

For now, Nuget manager just parses versions from ranges, stripping out any brackets if such an update would be valid.
It would be better to add support for ranges and their update strategies as well.

**Did you already have any implementation ideas?**

It's very similar to Maven ranges.

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**

No

**Is this a feature you'd be interested in implementing yourself?**

Yes

### PR DESCRIPTION
**Pull request by zharinov**
Closes #5016

### CODE REVIEW

**Comment by zharinov** [comment]
The confusing part is that `1.2.3` actually means `1.2.3` and above (but lower than `1.3.0`, I guess), while exact versions are like `[1.2.3]`.

File: lib/versioning/nuget/index.spec.ts [hunk 1]
```diff
@@ -1,60 +1,181 @@
 import nuget from '.';
```
Comment (zharinov) [inline lib/versioning/nuget/index.spec.ts:1]: @viceice @rarkins Before I proceed further, can you please take a look at these tests? Everything is correct?

File: lib/versioning/nuget/index.spec.ts [hunk 2]
```diff
@@ -1,60 +1,181 @@
...
-      '2.0.2-pre20191018090318',
-      '1.0.0+c30d7625',
-      '2.3.4-beta+1990ef74',
-    ].forEach((version) => {
-      it(version, () => {
-        expect(nuget.isStable(version)).toMatchSnapshot();
-      });
-    });
+  it('isStable', () => {
+    expect(nuget.isStable('9.0.3')).toBeTrue();
```
Comment (rarkins) [inline lib/versioning/nuget/index.spec.ts:28]: Didn't you say this represents a range, not a single version?

**Comment by rarkins** [comment]
> The confusing part is that `1.2.3` actually means `1.2.3` and higher (but lower than `1.3.0`, I guess), while exact versions are like `[1.2.3]`.

1. What's the best reference we can use regarding this?
2. Will all current use cases of nuget versioning need to behave this way, or do we need to split into two versioning schemes where `1.2.3` means an exact version or it means a range?

File: lib/versioning/nuget/index.spec.ts (hunk 2 above)
Comment (zharinov) [reply lib/versioning/nuget/index.spec.ts:28]: Yes, for now I'm confused with this aspect since for we currently treat these values as single versions.

Comment (rarkins) [reply lib/versioning/nuget/index.spec.ts:28]: I don't think there is a need for us to support functions like `isStable()` for strings for which `isSingleVersion()` returns false

**Comment by JamieMagee** [comment]
> 1. What's the best reference we can use regarding this?

https://docs.microsoft.com/en-us/nuget/concepts/package-versioning#version-ranges

> 2. Will all current use cases of nuget versioning need to behave this way, or do we need to split into two versioning schemes where `1.2.3` means an exact version or it means a range?

`1.2.3` has always meant a range. I don't think we should start pinning dependencies.

Quite frankly, NuGet versioning is rather confusing and a bit broken. It's not just me either, Jon Skeet discussed it on [.NET Rocks](https://dotnetrocks.com/?show=1684) and argued the whole ecosystem keeps going because major versions don't really break anything.

[Here's the doc on the dependency resolution algorithm](https://docs.microsoft.com/en-us/nuget/concepts/dependency-resolution) for NuGet if you want the reference.

**Comment by rarkins** [comment]
> `1.2.3` has always meant a range.

In that case we should consider it like a bug fix. I guess it also means in the past that if we were updating `1.2.3` to `1.2.4` then we thought we were doing a pinned version upgrade but in fact were bumping.

> I don't think we should start pinning dependencies.

Any reason why not to, if we have a way to detect "applications" or if users opt-in? The rationale for pinning seems pretty cross-language to me.

> Quite frankly, NuGet versioning is rather confusing and a bit broken. It's not just me either, Jon Skeet discussed it on [.NET Rocks](https://dotnetrocks.com/?show=1684) and argued the whole ecosystem keeps going because major versions don't really break anything.

I can imagine. Open ended ranges across major boundaries!

On this topic, doesn't NuGet resolve to the *minimum* matching in a range, instead of the maximum like most other languages?

**Comment by JamieMagee** [comment]
> Any reason why not to, if we have a way to detect "applications" or if users opt-in? The rationale for pinning seems pretty cross-language to me.

NuGet requires a single version of dependencies. If we start pinning versions NuGet might not be able to find a version that satisfies all the different ranges.

I think Maven also has this behaviour, but Maven allows you to exclude dependencies, in order to get a single version, but NuGet will just throw an exception.

**Comment by rarkins** [comment]
Thanks, good to know. That would also be an argument against bumping ranges too (which essentially achieves the same thing of only one version being allowed).

In that case we should permanently default auto to replace but maybe still allow pinning if users choose?

**Comment by JamieMagee** [comment]
I think we should put together a repo with some of these scenarios, and figure out what's useful and what breaks.

**Comment by viceice** [comment]
a good sample packages are [JsonSubTypes](https://www.nuget.org/packages/JsonSubTypes/1.8.0) v1.8.0 which depends on [Newtonsoft.Json](https://www.nuget.org/packages/Newtonsoft.Json/10.0.1) v10.0.1.

JsonSubTypes is compatible with >=10.0.1, so you can use Newtonsoft.Json v12.

Always works for compiletime, but can break on runtime. (only on net35 and net4x if assembly signing is used, workaround would be assembly binding redirection)

**Comment by rarkins** [comment]
I'm not sure what we should do with that information though

**Comment by viceice** [comment]
> I'm not sure what we should do with that information though

If we would pin the versions, nuget would not able to resolve a working tree

**Comment by rarkins** [comment]
Should we start by handing `1.2.3` correctly in a separate PR? e.g. `isSingleVersion('1.2.3') === false` because it's equivalent to `>=1.2.3` in `semver` logic? We could perhaps ensure that `rangeStrategy=bump` works for such cases so that people can configure it if they so wish.

**Comment by fgreinacher** [comment]
FYI @sieukrem Follow-up of https://github.com/renovatebot/renovate/pull/4947 Maybe you want to have a look 👍🏽

**Comment by sieukrem** [comment]
> FYI @sieukrem Follow-up of #4947 Maybe you want to have a look 👍🏽

I'm glad to see the discussion and thoughts done in same direction we did in #4947!

**Comment by JamieMagee** [comment]
I threw together [a sample repo](https://github.com/JamieMagee/RenovateNuGetTests) to test out some of the scenarios. [Check out the actions](https://github.com/JamieMagee/RenovateNuGetTests/actions?query=workflow%3Abuild) to see the results.

**Comment by rarkins** [comment]
Seems we need this to work:

```
isSingleVersion('1.2.3') === false
isSingleVersion('[1.2.3]') === true
```

**Comment by zharinov** [comment]
I wonder which is better:
```
isVersion('1.2.3') === false
isVersion('[1.2.3]') === true
```
or
```
isVersion('1.2.3') === false
isVersion('[1.2.3]') === true
```

The latter looks more correct, but means we have to (1) return `[1.2.3]` from `nuget` datasources and also (2) force `nuget` versioning (otherwise, all releases would be filtered out).

**Comment by rarkins** [comment]
image

**Comment by rarkins** [comment]
We use `isVersion()` to filter datasources, so I think we still need `isVersion('1.2.3')`
image
 to return `true` and `isVersion('[1.2.3]')` to return `false`.

We'll need to change a few things at once here. First of all, we need the manager to extract the full value e.g. with range syntax instead of just versions.

However `isSingleVersion('1.2.3')` should return `false` and `isSingleVersion(['1.2.3'])` could return `true`. We'd need to see if these existing concepts are enough or if we need to add new concepts to versioning's API because of nuget.

We also need to think about whether nuget's concept of "take the lowest matching version, not highest" needs us to change other logic. I suspect we may need to extend versioning APIs with a function that takes a constraint (e.g. version or range) and a list of `versions` and returns "the best match", instead of us assuming it's always the highest match.

**Comment by rarkins** [comment]
The closest concept to this I've seen before is in Python where `1.2.3` is a version, but `==1.2.3` is a "single version". With this one it has always been confusing, although we solved nearly all use cases (I think).

**Comment by zharinov** [comment]
- [ ] Extract full-qualified version values (separate PR?)
- [ ] Ensure these conditions are met:

    ```typescript
    isSingleVersion('1.2.3') === false
    isSingleVersion('[1.2.3]') === true
    ```

- [ ] Do we need versioning API changes?
- [ ] Incorporate the concept of *"take the lowest matching version, not highest"*.
  - Do we need API changes? I.e. best match instead of highest one.

**Comment by rarkins** [comment]
> * [ ]  Extract full-qualified version values (separate PR?)

I'm not sure that we can easily separate parts of this PR like that without having a broken state in-between. e.g. if we extract the full constraint string then it then won't parse our versioning.isValid()?

> * [ ]  Do we need versioning API changes?
> * [ ]  Incorporate the concept of _"take the lowest matching version, not highest"_.
>   
>   * Do we need API changes? I.e. best match instead of highest one.

I think these are related. Here is the line where we assume the maximum satisfying version is the one we want: https://github.com/renovatebot/renovate/blob/630b75b3e24042ba3386d44588becb62178c8540/lib/workers/repository/process/lookup/index.ts#L114

We could perhaps rename that from `maxSatisfyingVersion` to `getSatisfyingVersion` and then implement nuget versioning to return the lowest matching instead of highest.

**Comment by viceice** [comment]
I'm not sure, why we want the lowest possible update here? Wouldn't this cause  multiple updates?

Eg we have `1.2.3` in package file and `1.2.4`,`1.2.5`,`1.2.6`,`1.3.0` available. (`seperateMinorPatch=true`)
Now my expectation is that renovate will open a pr for `1.2.6` and `1.3.0`.

But if we implement *"take the lowest matching version, not highest"* it would open `1.2.4` and `1.3.0` ? So with  automerge patch we would cascade every single patch update instead of sugesting latest patch only?

**Comment by rarkins** [comment]
1.2.3 in a package file means >=1.2.3 doesn't it?

**Comment by viceice** [comment]
Yes, it means a restore will try to resolve lowest possible version by default.
I've configured the oposite on some of my repos, so a restore will resolve the highest possible version.

Those handling is ok for restore, but renovate should ignore this and always suggest latest versions as currently done.

**Comment by rarkins** [comment]
See https://github.com/renovatebot/renovate/pull/7917#issuecomment-742363793

**Comment by viceice** [comment]
OK, so it makes sense to use the lowest matching version now, but we need to be able to optin to highest, if configured in `nuget.config`

**Comment by rarkins** [comment]
@viceice can you make a feature request for that? We need to work out how to pass that elegantly to versioning

**Comment by rarkins** [comment]
I just realized that we already had both `getMaxSatisfyingVersion` and `getMinSatisfyingVersion` so that makes the most recent rename seem wrong.

Do we even need min/max? Could we just use `satisfies`, `sort` and `pop`?

**Comment by rarkins** [comment]
Let's close this for now, but keep it as a base to work off if someone with a desire to get this working wants to pick it up.

**Pull request closed without merging**